tests/test_cice_grid.py

```diff
@@ -0,0 +1,197 @@
+
+        # generate an tripolar grid as test data
+        run(
+            [
+                "ocean_grid_generator.py",
+                "-r",
+                "0.25",  # 4 degree grid
+                "--no_south_cap",
+                "--ensure_nj_even",
+                "-f",
+                self.path,
+            ]
+        )
+
+        # open ocean_hgrid.nc
+        self.ds = xr.open_dataset(self.path)
+
+        # an ocean mask with a arbitrary mask
+        self.mask_ds = xr.Dataset()
+        self.mask_ds["mask"] = (self.ds.area.coarsen(ny=2).sum().coarsen(nx=2).sum()) > 5e9
+        self.mask_ds.to_netcdf(self.mask_path)
+
+
+class CiceGrid:
+    """Make the CICE grid, using script under test"""
+
+    def __init__(self, mom_grid, tmp_path):
+        self.path = str(tmp_path) + "/grid.nc"
+        self.kmt_path = str(tmp_path) + "/kmt.nc"
+        cice_grid = CiceGridNc(self.path, self.kmt_path)
+        cice_grid.build_from_mom(mom_grid.path, mom_grid.mask_path)
+        self.ds = xr.open_dataset(self.path, decode_cf=False)
+        self.kmt_ds = xr.open_dataset(self.kmt_path, decode_cf=False)
+
+
+# pytest doesn't support class fixtures, so we need these two constructor funcs
+@pytest.fixture
+def mom_grid(tmp_path):
+    return MomGrid(tmp_path)
+
+
+@pytest.fixture
+def cice_grid(mom_grid, tmp_path):
+    return CiceGrid(mom_grid, tmp_path)
+
+
+@pytest.fixture
+def test_grid_ds(mom_grid):
+    # this generates the expected answers
+
+    ds = mom_grid.ds
+
+    test_grid = xr.Dataset()
+    # corners of the cice grid are NE corner
+    test_grid["ulat"] = deg2rad(ds.y.isel(nxp=slice(2, None, 2), nyp=slice(2, None, 2)))
+    test_grid["ulon"] = deg2rad(ds.x.isel(nxp=slice(2, None, 2), nyp=slice(2, None, 2)))
+
+    # centers of cice grid
+    test_grid["tlat"] = deg2rad(ds.y.isel(nxp=slice(1, None, 2), nyp=slice(1, None, 2)))
+    test_grid["tlon"] = deg2rad(ds.x.isel(nxp=slice(1, None, 2), nyp=slice(1, None, 2)))
+
+    # length of top edge of cells
+    test_grid["htn"] = ds.dx.isel(nyp=slice(2, None, 2)).coarsen(nx=2).sum() * 100
+    # length of right edge of cells
+    test_grid["hte"] = ds.dy.isel(nxp=slice(2, None, 2)).coarsen(ny=2).sum() * 100
+
+    # angle at u point
+    test_grid["angle"] = deg2rad(ds.angle_dx.isel(nyp=slice(2, None, 2), nxp=slice(2, None, 2)))
+    # angle a t points
+    test_grid["angleT"] = deg2rad(ds.angle_dx.isel(nyp=slice(1, None, 2), nxp=slice(1, None, 2)))
+
+    # area of cells
+    test_grid["tarea"] = mom_grid.ds.area.coarsen(ny=2).sum().coarsen(nx=2).sum()
+
+    # uarea is area of a cell centred around the u point
+    # we need to wrap in latitude and fold on longitude to calculate this
+    area_wrapped = mom_grid.ds.area
+    area_wrapped = xr.concat([ds.area.isel(nx=slice(1, None)), ds.area.isel(nx=0)], dim="nx")
+
+    top_row = xr.concat([ds.area.isel(ny=-1, nx=slice(-2, 0, -1)), ds.area.isel(ny=-1, nx=[-1, 0])], dim="nx")
```

Comment:
```suggestion
    top_row = area_wrapped.isel(ny=-1, nx=slice(-1, 0, -1))
```
Do we just want a reversed copy of the wrapped top row here, or have I misunderstood?
(Or perhaps the folded top row should be concatenated before wrapping? The order of these operations affects the index of the tripole singularities - see fig 9.9 of Griffies - note that the singularities fall on u points 0 and ni/2)
![Screenshot 2024-04-05 at 6 28 09 pm](https://github.com/COSIMA/om3-utils/assets/31054815/d26941d3-c6d2-45f3-9722-13d94a486adc)

Reply:
I think I tried this suggestion and the resulting areas were offset by 1 cell ... I will have to investigate. I am confident the areas from this method are right, but maybe the method is wrong.

Reply:
Ah ...

xarray is confusing sometimes. 

`ds.area.isel(ny=-1, nx=[-1, 0])` and `ds.area.isel(ny=-1, nx=[0,-1])` are equal :)

Anyway the slightly better version of this line would be:

`xr.concat([ds.area.isel(ny=-1, nx=slice(-2, None, -1)), ds.area.isel(ny=-1, nx=-1)], dim="nx")`

But the even better version, is to fold first, then wrap, and then its generally clearer in total!

Reply:
Changed to fold first, then wrap